pkg/test/framework/testcontext.go

```diff
@@ -0,0 +1,301 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package framework
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"reflect"
+	"testing"
+
+	"istio.io/istio/pkg/test/framework/components/environment"
+	"istio.io/istio/pkg/test/framework/core"
+	"istio.io/istio/pkg/test/framework/label"
+	"istio.io/istio/pkg/test/framework/resource"
+	"istio.io/istio/pkg/test/scopes"
+)
+
+// TestContext is a test-level context that can be created as part of test executing tests.
+type TestContext interface {
+	resource.Context
+
+	// GoTest (Deprecated: this will be removed soon!) provides access to the underlying *testing.T.
+	GoTest() *testing.T
+
+	// NewSubTest creates a new sub-test under the current running Test. The lifecycle of a sub-Test is scoped to the
+	// parent. Calls to Done() will block until all children are also Done(). When Run, sub-Tests will automatically
+	// create their own Golang *testing.T with the name provided.
+	//
+	// If this TestContext was not created by a Test or if that Test is not running, this method will panic.
+	NewSubTest(name string) *Test
```

Comment:
Consider adding a "Run" method here as well. I suspect most sub-tests will inherit its parent facets, so it will probably be quite convenient to simply do ctx.Run(...)

Reply:
Isn't to goal to have folks use the `Test` API for running? If so, I think I'd rather not even expose that here.

Reply:
Test is useful for doing Requires or Label operations. If you don't need those, then short-circuiting to "Run" should be fine. This is more prevalent at the top-level main, as it is quite likely that you will need to run some setup function (hence framework.Main/Run would be seldom used). In subtests, it is more likely that the sub test will inherit the parent test's traits.

Not cirticial for this check-in though, we can add it later.